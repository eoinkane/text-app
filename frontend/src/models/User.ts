import axios, { AxiosResponse } from "axios";

interface UserProps {
  firstName?: string;
  lastName?: string;
  id?: number;
}

type valueof<T> = T[keyof T];

export default class User {
  constructor(private data: UserProps) {}

  get<K extends keyof UserProps>(propName: K): valueof<UserProps> {
    return this.data[propName];
  }

  set(update: UserProps): void {
    Object.assign(this.data, update);
  }
  fetch(): void {
    // axios
    //   .get(`http://localhost:3000/users/${this.get("id")}`)
    //   .then((response: AxiosResponse): void => {
    //     this.set(response.data);
    //   });
  }

  save(): void {
    // const id = this.get("id");
    // if (id) {
    //   axios.put(`http://localhost:3000/users/${id}`, this.data);
    // } else {
    //   axios.post(`http://localhost:3000/users`, this.data);
    // }
  }
}
