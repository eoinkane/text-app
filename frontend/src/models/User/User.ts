import axios, { AxiosResponse, AxiosError } from "axios";

interface UserProps {
  username: string;
  firstName: string;
  lastName: string;
  id: number;
}

type valueof<T> = T[keyof T];

export class CreateError extends Error {}

export type CreatePayload = {
  username: string;
  password: string;
  firstName: string;
  lastName: string;
};

export default class User {
  private data: UserProps | undefined;
  constructor() {}

  get<K extends keyof UserProps>(propName: K): valueof<UserProps> {
    if (!this.data) {
      throw new Error("User is undefined");
    }
    return this.data[propName];
  }

  private set(update: UserProps): void {
    this.data = !this.data ? update : Object.assign(this.data, update);
  }

  init(data: UserProps): void {
    this.set(data);
  }

  static async create(payload: CreatePayload): Promise<User> {
    let data: AxiosResponse | AxiosError;
    try {
      const response = await axios.post(
        "http://127.0.0.1:5000/api/users",
        {
          firstName: payload.firstName,
          lastName: payload.lastName,
        },
        {
          auth: {
            username: payload.username,
            password: payload.password,
          },
        }
      );
      data = response;

      const newUser = new User();

      newUser.init(data.data);
      return newUser;
    } catch (err) {
      console.log(err);
      throw new CreateError("Error creating new User");
    }
  }

  // save(): void {
  //   const id = this.get("id");
  //   if (id) {
  //     axios.put(`http://localhost:3000/users/${id}`, this.data);
  //   } else {
  //     axios.post(`http://localhost:3000/users`, this.data);
  //   }
  // }
}
