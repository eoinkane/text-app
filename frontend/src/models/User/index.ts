import type { CreatePayload } from "./User";
export { default } from "./User";
export { CreateError } from "./User";
export type { CreatePayload };
