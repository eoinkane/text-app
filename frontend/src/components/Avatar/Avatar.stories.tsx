import React from "react";
import User from "../../models/User/";
import AvatarComponent from "./index";

export default {
  title: "Avatar",
  component: AvatarComponent,
};

export const AvatarWithFullUser = () => {
  const user = new User();
  user.init({
    id: 0,
    username: "test",
    firstName: "Jane",
    lastName: "Doe",
  });
  return <AvatarComponent user={user} />;
};
