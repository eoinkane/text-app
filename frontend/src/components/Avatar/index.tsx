import React, { Fragment } from "react";
import User from "../../models/User/";
import { Avatar as MaterialUIAvatar } from "@material-ui/core";

interface IAvatarProps {
  user: User;
}

const Avatar: React.FC<IAvatarProps> = ({ user }) => {
  const firstName = user.get("firstName") as string;
  const lastName = user.get("lastName") as string;
  const initals =
    !firstName || !lastName ? null : `${firstName[0]}${lastName[0]}`;
  return (
    <Fragment>
      <MaterialUIAvatar>{initals}</MaterialUIAvatar>
    </Fragment>
  );
};

export default Avatar;
