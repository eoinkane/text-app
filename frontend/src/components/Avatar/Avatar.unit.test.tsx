import React from "react";
import { shallow } from "enzyme";
import User from "./../../models/User";
import Avatar from "./index";

import MaterialUIAvatar from "@material-ui/core/Avatar";

describe("Avatar", () => {
  it("Renders", () => {
    const user = new User({ firstName: "Jane", lastName: "Doe" });
    const avatar = shallow(<Avatar user={user} />);
    expect(avatar).toMatchSnapshot();
  });

  it("Renders with correct text when first name & last name are passed", () => {
    const user = new User({ firstName: "Jane", lastName: "Doe" });
    const firstName = user.get("firstName") as string;
    const lastName = user.get("lastName") as string;
    const initials = `${firstName[0]}${lastName[0]}`;
    const avatar = shallow(<Avatar user={user} />);

    expect(avatar).toMatchSnapshot();
    expect(avatar.find(MaterialUIAvatar).text()).toEqual(initials);
  });
});
