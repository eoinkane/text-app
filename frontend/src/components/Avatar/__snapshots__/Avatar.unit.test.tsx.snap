// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Renders 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Avatar))>
    JD
  </WithStyles(ForwardRef(Avatar))>
</Fragment>
`;

exports[`Avatar Renders with correct text when first name & last name are passed 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Avatar))>
    JD
  </WithStyles(ForwardRef(Avatar))>
</Fragment>
`;
